### PR TITLE
Rule S3655: Update SE to the new CFG and add C# 8 unit tests EmptyNullableValueAccess

### LINKED ISSUE #2893: Rule S3655: Update SE to the new CFG and add C# 8 unit tests EmptyNullableValueAccess
Subtask of #2849

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #2893

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyNullableValueAccess.cs [hunk 1]
```diff
@@ -116,6 +116,35 @@ public void TestComplexCondition(int? i3, double? d3, float? f3)
                 Console.WriteLine(f3.Value); // TODO: Should be NC
             }
         }
+
+        public int CSharp8_SwitchExpressions(bool zero)
+        {
+            int? i = zero switch { true => 0, _ => null };
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyNullableValueAccess.cs]: We can add one more case with noncompliant access inside switch. What do you think?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyNullableValueAccess.cs [hunk 2]
```diff
@@ -116,6 +116,50 @@ public void TestComplexCondition(int? i3, double? d3, float? f3)
...
+        }
+
+        public int CSharp8_SwitchExpressions3(int? value)
+        {
+            return value.HasValue switch { true => value.Value, false => 0};
+        }
+
+        public int CSharp8_SwitchExpressions4(int? value)
+        {
+            return value.HasValue switch { true => 0, false => value.Value };   // FN - switch expressions are not constrained
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyNullableValueAccess.cs]: Would be great to add an issue for this (and a comment with the link) to be sure that we don't forget about it.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyNullableValueAccess.cs [hunk 3]
```diff
@@ -141,4 +185,14 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
         }
     }
 
+    public interface IWithDefaultImplementation
+    {
+        //Default interface methods
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyNullableValueAccess.cs]: pedantic: the comment is a bit redundant and can be removed

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyNullableValueAccess.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyNullableValueAccess.cs]: I've created https://github.com/SonarSource/sonar-dotnet/issues/2949 without backlink, there are more rules affected.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyNullableValueAccess.cs [hunk 4]
```diff
@@ -116,6 +116,50 @@ public void TestComplexCondition(int? i3, double? d3, float? f3)
...
+                return intOrNull.Value; // FN - content of static local function is not inspected by SE
+            }
+
+            return ExtractValue(param);
+        }
+
+        public int CSharp8_NullCoalescingAssignment(int? param)
+        {
+            param ??= 42;
+            return param.Value; // OK, value is allways set
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EmptyNullableValueAccess.cs]:
```suggestion
            return param.Value; // OK, value is always set
```

**Pull request merged**